src/linear_algebra/invariant_submodule.lean

```diff
@@ -0,0 +1,142 @@
+
+/-- `U` is `T` invariant is equivalent to saying `T(U) ⊆ U` -/
+lemma invariant_under_iff (U : submodule R E) (T : E →ₗ[R] E) :
+  U.invariant_under T ↔ T '' U ⊆ U := by rw [← set.maps_to', U.invariant_under_iff_maps_to]
+
+variables (U V : submodule R E) (hUV : is_compl U V) (T : E →ₗ[R] E)
+
+local notation `pᵤ` := submodule.linear_proj_of_is_compl U V hUV
+local notation `pᵥ` := submodule.linear_proj_of_is_compl V U hUV.symm
+
+lemma proj_comp_self_comp_proj_eq_of_invariant_under
+  (h : U.invariant_under T) (x : E) : ↑(pᵤ (T ↑(pᵤ x))) = T ↑(pᵤ x) :=
+begin
+  rw submodule.linear_proj_of_is_compl_eq_self_iff,
+  exact h (submodule.coe_mem _),
+end
+
+lemma invariant_under_of_proj_comp_self_comp_proj_eq
+  (h : ∀ x : E, ↑(pᵤ (T ↑(pᵤ x))) = T ↑(pᵤ x)) : U.invariant_under T :=
+begin
+  intros u hu,
+  rw [submodule.mem_comap, ← submodule.linear_proj_of_is_compl_eq_self_iff hUV _,
+      ← (submodule.linear_proj_of_is_compl_eq_self_iff hUV u).mpr hu, h],
+end
+
+lemma proj_comp_self_comp_proj_eq_iff_invariant_under :
+  U.invariant_under T ↔ (∀ x : E, ↑(pᵤ (T ↑(pᵤ x))) = T ↑(pᵤ x)) :=
+⟨U.proj_comp_self_comp_proj_eq_of_invariant_under V hUV T,
+ U.invariant_under_of_proj_comp_self_comp_proj_eq V hUV T⟩
+
+lemma proj_comp_self_comp_proj_eq_iff_invariant_under' :
+  V.invariant_under T ↔ (∀ x : E, (pᵤ (T ↑(pᵤ x)) : E) = pᵤ (T x)) :=
+by simp_rw [submodule.proj_comp_self_comp_proj_eq_iff_invariant_under _ _ hUV.symm,
+            linear_proj_of_is_compl_eq_self_sub_linear_proj, map_sub, sub_eq_self,
+            submodule.coe_sub, sub_eq_zero, eq_comm]
+
+lemma compl_invariant_under_iff_linear_proj_and_T_commute :
+  (U.invariant_under T ∧ V.invariant_under T) ↔ commute (U.subtype.comp (pᵤ)) T :=
+begin
+  simp_rw [commute, semiconj_by, linear_map.ext_iff, linear_map.mul_apply,
+           linear_map.comp_apply, U.subtype_apply],
+  split,
+  { rintros ⟨h1, h2⟩ x,
+    rw [← (U.proj_comp_self_comp_proj_eq_iff_invariant_under' V _ _).mp h2 x],
+    exact (submodule.linear_proj_of_is_compl_eq_self_iff hUV _).mpr
+      (h1 (set_like.coe_mem (pᵤ x))) },
+  { intros h,
+    split,
+    { simp_rw [U.proj_comp_self_comp_proj_eq_iff_invariant_under _ hUV, h,
+               submodule.linear_proj_of_is_compl_idempotent hUV],
+      exact λ x, rfl },
+    { simp_rw [U.proj_comp_self_comp_proj_eq_iff_invariant_under' _ hUV, h,
+               submodule.linear_proj_of_is_compl_idempotent hUV],
+      exact λ x, rfl } }
+end
+
+lemma commutes_with_linear_proj_iff_linear_proj_eq [invertible T] :
+  commute (U.subtype.comp pᵤ) T ↔
+    (⅟ T).comp ((U.subtype.comp pᵤ).comp T) = U.subtype.comp pᵤ :=
+begin
+  rw [commute, semiconj_by],
+  simp_rw [linear_map.mul_eq_comp],
+  split,
+  { intro h,
+    simp_rw [h, ← linear_map.mul_eq_comp, ← mul_assoc, inv_of_mul_self, one_mul], },
+  { intros h,
+    rw ← h, simp_rw [← linear_map.mul_eq_comp, ← mul_assoc, mul_inv_of_self],
+    rw [mul_assoc (⅟ T) _ _],
+    simp_rw [linear_map.mul_eq_comp, h], refl, }
+end
```

Comment:
I feel like this would be more useful as a theorem about `linear_equiv` (especially because we don't have that much API for `inv_of`). I'm not 100% sure about it though, so if you (or Eric) really think that `invertible` is better I'd be happy to hear why and change my mind.

In this case, you can greatly shorten your proof by using `linear_equiv.eq_to_linear_map_symm_comp` (I can't find a similar lemma for `inv_of`), for example:

```suggestion
lemma commutes_with_linear_proj_iff_linear_proj_eq {S : E ≃ₗ[R] E} :
  commute (U.subtype.comp pᵤ) S ↔
    (S.symm : E →ₗ[R] E) ∘ₗ ((U.subtype.comp pᵤ).comp (S : E →ₗ[R] E)) = U.subtype.comp pᵤ :=
begin
  rw [commute, semiconj_by],
  simp_rw [← linear_equiv.to_linear_map_eq_coe, linear_map.mul_eq_comp],
  rw [eq_comm, ← linear_equiv.eq_to_linear_map_symm_comp, eq_comm]
end
```

Reply:
how about I add a lemma showing that there exists a linear_equiv map that is equal to the invertible linear map? so that we can use the linear_equiv results, but also keep the invertible stuff (because I personally think it's clearer in this context)

Reply:
not sure where to put them though